[FLINK-3703][cep] Add sequence matching semantics to discard matched events

mushketyk opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.
- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

This PR adds sequence matching semantics to matched events, so a user can specify if a particular event can be used in multiple matching sequences. 
I didn't add user documentation yet, but I'll add it when someone from Flink community will check if this PR implements matching semantics correctly.

smarthi commented on flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java:
```diff
@@ -26,6 +26,7 @@
 import akka.pattern.Patterns;
 import akka.util.Timeout;
 
+import com.google.common.base.Joiner;
```
Avoid using Guava API calls.  You could use Commons lang StringUtils.join() instead.

mushketyk replied:
Good point. Will update.

smarthi commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java:
```diff
@@ -256,16 +267,75 @@ public void prune(long pruningTimestamp) {
 	}
 
+	private void doCleanUp(Predicate<K, V> predicate) {
+		ArrayList<SharedBufferEntry<K, V>> toRemove = new ArrayList<>();
```
Any reason for using a Concreate class - ArrayList on the LS here ? Can it be 
`List<SharedBufferEntry<K, V>> toRemove = new ArrayList<>();`

smarthi commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java:
```diff
@@ -256,16 +267,75 @@ public void prune(long pruningTimestamp) {
+		}
+	}
+
+	private boolean doRecursiveCleanup(SharedBufferEntry<K, V> startNode, Predicate<K, V> cleanUp, ArrayList<SharedBufferEntry<K, V>> toRemove) {
```
Replace ArrayList by List in the arguments, unless we need it to be ArrayList explicitly.

tillrohrmann commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java:
```diff
@@ -243,8 +252,7 @@ public int hashCode() {
 				try {
-					if (stateTransition.getCondition() == null || stateTransition.getCondition().filter(event)) {
-						// filter condition is true
+					if (stateTransition.isConditionTrue(event)) {
```
Good refactoring

tillrohrmann commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java:
```diff
@@ -133,9 +134,13 @@ public void addState(final State<T> state) {
 		for (int i = 0; i < numberComputationStates; i++) {
+			if (skipOtherStates) {
+				continue;
+			}
```
I think you cannot simply skip the current set of computation states because then you would retain those states, which have a next condition and would thus be filtered out in the `MatchingBehaviour.AFTER_LAST` case.

tillrohrmann commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java:
```diff
@@ -300,17 +311,21 @@ public int hashCode() {
+										event,
+										timestamp,
+										newComputationStateVersion,
+										startTimestamp));
```
Ideally we would filtered out all computation states which referred to pruned computations. Then we would not have to make this case distinction here.

tillrohrmann commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java:
```diff
@@ -202,7 +203,8 @@ public void prune(long pruningTimestamp) {
 		final long timestamp,
-		final DeweyNumber version) {
+		final DeweyNumber version,
+		final MatchingBehaviour matchingBehaviour) {
```
I think the `SharedBuffer` should nothing know about the `MatchingBehaviour`. The `MatchingBehaviour` is something specific to the `NFA` and, thus, the `NFA` should be responsible for implementing it. This will give a better separation of concerns.

tillrohrmann commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/State.java:
```diff
@@ -38,10 +40,16 @@
 	private final String name;
 	private final StateType stateType;
 	private final Collection<StateTransition<T>> stateTransitions;
+	private final MatchingBehaviour matchingBehaviour;
```
The `State` shouldn't have to know the `MatchingBehaviour`. They should be agnostic to the matching behaviour.

tillrohrmann commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/Pattern.java:
```diff
@@ -129,6 +141,23 @@ public Time getWindowTime() {
+			pattern = (Pattern<T, F>) pattern.getPrevious();
+		}
+
+		return this;
```
Looking at the necessary implementation to wire the matching behaviour through all `Pattern` instances, I wonder whether the `MatchingBehaviour` should rather be given to the `CEP.pattern` call. There it could be given to the generated `NFA` which is responsible for making sure that the behaviour is respected. The API would then look like:

`CEP.pattern(input, pattern, matchingBehaviour)`

What do you think?

tillrohrmann commented:
Thanks for your contribution and apologies for the late review @mushketyk. I think your implementation goes into the right direction and I really like the testing.

I think what we still can and should improve is to pull out the `MatchingBehaviour` out of the `SharedBuffer` and the `State`. The `NFA` should be the only component which is responsible for the matching behaviour. I think we can achieve this if we pimp the `SharedBuffer` a little bit.

So what should roughly happen is that the `NFA` removes computation states which are not longer valid. Furthermore, we have to clean up the `SharedBuffer` and remove `SharedBufferEntries` which depend on other `SharedBufferEntries` which are no longer valid (e.g. due to `MatchingBehaviour.afterLast`). 

So when you have a `SharedBufferEntry` representing a certain state which is part of the currently matched sequence and `MatchingBehaviour.afterLast`, then you have to find out all `SharedBufferEntries` which depend on this state (children) and remove them. That you have to do until you reach the youngest descendant. Currently we can only find the predecessors for a given `SharedBufferEntry` but not the children. I think that would have to be added in order to implement the different `MatchingBehaviours`.

mushketyk commented:
@tillrohrmann Thank you for the very detailed review. I do appreciate it.
I agree with your suggestions and I'll try to improve the code and write documentation in the next couple of days.
At the very least I know that tests are correct, so I can start changing the implementation.

mushketyk commented:
Hi @tillrohrmann 
I've updated the PR according to your review. I had to add reverse reverse edges to buffer entries and changed the interface of the SharedBuffer a bit to move code from it to the NFA class. I hope I didn't break the encapsulation too much, so I would to see your opinion about this.
I also added more tests and user documentation in this PR.

LordFB commented:
Hi @mushketyk and @tillrohrmann,

are there any updates on this Pull Request? Or is there already a way to change the matching behaviour in FlinkCEP?

mushketyk commented:
Hi @LordFB 

I don't think there is currently a way to do it in cep-flink. I am still waiting for Till's review for this PR, but he seems to be really busy with other work.

Maybe together we will be able to convince him to spend some time on reviewing this and similar CEP PRs.

LordFB commented:
Hi @mushketyk,

too bad, with this missing it is kind of a gamebreaker for Flink in my Use Case.

Jeah, that'd be great, if Till would show some action on this and the related PRs.

kl0u commented:
Hi @mushketyk . 

I am trying to clean up a bit the open PRs for the CEP library and the related JIRAs. 
The issue that this PR covers was already implemented as part of multiple other features that I also mention in the related JIRA. 
Could you close this PR?

mushketyk commented:
@kl0u Sure.
